controlplane/kubeadm/internal/workload_cluster_conditions.go

```diff
@@ -0,0 +1,531 @@
+// This operation is best effort, in the sense that in case of problems in retrieving member status, it sets
+// the condition to Unknown state without returning any error.
+func (w *Workload) UpdateEtcdConditions(ctx context.Context, kcp *controlplanev1.KubeadmControlPlane, machines []*clusterv1.Machine) {
+	isEtcdManaged := kcp.Spec.KubeadmConfigSpec.ClusterConfiguration == nil || kcp.Spec.KubeadmConfigSpec.ClusterConfiguration.Etcd.External == nil
+	if isEtcdManaged {
+		w.updateStakedEtcdConditions(ctx, kcp, machines)
+		return
+	}
+	w.updateExternalEtcdConditions(ctx, kcp, machines)
+}
+
+func (w *Workload) updateExternalEtcdConditions(ctx context.Context, kcp *controlplanev1.KubeadmControlPlane, machines []*clusterv1.Machine) { //nolint:unparam
+	// When KCP is not responsible for external etcd, we are reporting only health at KCP level.
+	conditions.MarkTrue(kcp, controlplanev1.EtcdClusterHealthyCondition)
+
+	// TODO: check external etcd for alarms an possibly also for member errors
+	// this requires implementing an new type of etcd client generator given that it is not possible to use nodes
+	// as a source for the etcd endpoint address; the address of the external etcd should be available on the kubeadm configuration.
+}
+
+func (w *Workload) updateStakedEtcdConditions(ctx context.Context, kcp *controlplanev1.KubeadmControlPlane, machines []*clusterv1.Machine) {
+	// NOTE: This methods uses control plane nodes only to get in contact with etcd but then it relies on etcd
+	// as ultimate source of truth for the list of members and for their health.
+	controlPlaneNodes, err := w.getControlPlaneNodes(ctx)
+	if err != nil {
+		conditions.MarkUnknown(kcp, controlplanev1.EtcdClusterHealthyCondition, controlplanev1.EtcdClusterInspectionFailedReason, "Failed to list nodes which are hosting the etcd members")
+		for _, m := range machines {
+			conditions.MarkUnknown(m, controlplanev1.MachineEtcdMemberHealthyCondition, controlplanev1.EtcdMemberInspectionFailedReason, "Failed to get the node which is hosting the etcd member")
+		}
+		return
+	}
+
+	// Update conditions for etcd members on the nodes.
+	var additionalKCPErrors []string
+	var clusterID *uint64
+	var members []*etcd.Member
+
+	for _, node := range controlPlaneNodes.Items {
+		// Search for the machine corresponding to the node.
+		var machine *clusterv1.Machine
+		for _, m := range machines {
+			if m.Status.NodeRef != nil && m.Status.NodeRef.Name == node.Name {
+				machine = m
+			}
+		}
+
+		if machine == nil {
+			// If there are machines still provisioning there is the chance that the node will be linked to a machine soon,
+			// otherwise report the error at KCP level given that there is no machine to report on.
+			if hasProvisioningMachine(machines) {
+				continue
+			}
+			additionalKCPErrors = append(additionalKCPErrors, fmt.Sprintf("Control plane node %s does not have a corresponding machine", node.Name))
```

Comment:
Even if this is a remote chance, if someone adds a control-plane node not managed by KCP/CAPI (without a machine), this could be relevant e.g. when making assumptions about quorum for remediation.
Given that this is a remote chance, I'm personally -1 to add another condition and I prefer to merge this in an existing one (avoid verbosity).